src/jitlayers.cpp

```diff
@@ -197,47 +197,84 @@ static jl_callptr_t _jl_compile_codeinst(
     return fptr;
 }
 
-// get the address of a C-callable entry point for a function
+Function *jl_generate_ccallable(void *llvmmod, void *sysimg_handle, jl_value_t *declrt, jl_value_t *sigt, jl_codegen_params_t &params);
+
+// compile a C-callable alias
 extern "C" JL_DLLEXPORT
-void *jl_function_ptr(jl_function_t *f, jl_value_t *rt, jl_value_t *argt)
+void jl_compile_extern_c(void *llvmmod, void *p, void *sysimg, jl_value_t *declrt, jl_value_t *sigt)
 {
-    JL_GC_PUSH1(&argt);
     JL_LOCK(&codegen_lock);
     jl_codegen_params_t params;
-    Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt, params);
-    jl_jit_globals(params.globals);
-    assert(params.workqueue.empty());
-    jl_add_to_ee();
-    JL_GC_POP();
-    void *ptr = (void*)getAddressForFunction(llvmf->getName());
+    jl_codegen_params_t *pparams = (jl_codegen_params_t*)p;
+    if (pparams == NULL)
+        pparams = &params;
+    jl_generate_ccallable(llvmmod, sysimg, declrt, sigt, *pparams);
+    if (!sysimg) {
+        if (p == NULL) {
+            jl_jit_globals(params.globals);
+            assert(params.workqueue.empty());
+        }
+        if (llvmmod == NULL)
+            jl_add_to_ee();
+    }
     JL_UNLOCK(&codegen_lock);
-    return ptr;
 }
 
-// export a C-callable entry point for a function (dllexport'ed dlsym), with a given name
+bool jl_type_mappable_to_c(jl_value_t *ty);
+
+// declare a C-callable entry point; called during code loading from the toplevel
 extern "C" JL_DLLEXPORT
-void jl_extern_c(jl_function_t *f, jl_value_t *rt, jl_value_t *argt, char *name)
-{
+void jl_extern_c(jl_value_t *declrt, jl_tupletype_t *sigt)
+{
+    // validate arguments. try to do as many checks as possible here to avoid
+    // throwing errors later during codegen.
+    JL_TYPECHK(@ccallable, type, declrt);
```

Comment:
Since this function is named `ccallable`, I don't think the `@` is really a part of the name.

Reply:
The user-facing API is `@ccallable`. I'll add an underscore to `ccallable` to make that clear.